neural_modelling/src/neuron/implementations/neuron_impl_external_devices.h

```diff
@@ -0,0 +1,572 @@
+
+    // Get threshold and additional input parameters for this neuron
+    threshold_type_t *the_threshold_type = &threshold_type_array[neuron_index];
+    additional_input_t *additional_inputs =
+            &additional_input_array[neuron_index];
+    synapse_param_t *the_synapse_type =
+            &neuron_synapse_shaping_params[neuron_index];
+
+    // Store whether the neuron has spiked
+    bool has_spiked = false;
+
+    // Loop however many times requested; do this in reverse for efficiency,
+    // and because the index doesn't actually matter
+    for (uint32_t i = n_steps_per_timestep; i > 0; i--) {
+        // Get the voltage
+        state_t soma_voltage = neuron_model_get_membrane_voltage(this_neuron);
+
+        // Get the exc and inh values from the synapses
+        input_t *exc_values =
+                synapse_types_get_excitatory_input(the_synapse_type);
+        input_t *inh_values =
+                synapse_types_get_inhibitory_input(the_synapse_type);
+
+        // Call functions to obtain exc_input and inh_input
+        input_t *exc_input_values = input_type_get_input_value(
+                exc_values, input_types, NUM_EXCITATORY_RECEPTORS);
+        input_t *inh_input_values = input_type_get_input_value(
+                inh_values, input_types, NUM_INHIBITORY_RECEPTORS);
+
+        // Sum g_syn contributions from all receptors for recording
+        REAL total_exc = 0;
+        REAL total_inh = 0;
+
+        for (int i = 0; i < NUM_EXCITATORY_RECEPTORS; i++) {
+            total_exc += exc_input_values[i];
+        }
+        for (int i = 0; i < NUM_INHIBITORY_RECEPTORS; i++) {
+            total_inh += inh_input_values[i];
+        }
+
+        // Do recording if on the first step
+        if (i == n_steps_per_timestep) {
+            neuron_recording_record_accum(
+                    V_RECORDING_INDEX, neuron_index, soma_voltage);
+            neuron_recording_record_accum(
+                    GSYN_EXC_RECORDING_INDEX, neuron_index, total_exc);
+            neuron_recording_record_accum(
+                    GSYN_INH_RECORDING_INDEX, neuron_index, total_inh);
+        }
+
+        // Call functions to convert exc_input and inh_input to current
+        input_type_convert_excitatory_input_to_current(
+                exc_input_values, input_types, soma_voltage);
+        input_type_convert_inhibitory_input_to_current(
+                inh_input_values, input_types, soma_voltage);
+
+        external_bias += additional_input_get_input_value_as_current(
+                additional_inputs, soma_voltage);
+
+        // update neuron parameters
+        state_t result = neuron_model_state_update(
+                NUM_EXCITATORY_RECEPTORS, exc_input_values,
+                NUM_INHIBITORY_RECEPTORS, inh_input_values,
+                external_bias, this_neuron);
+
+        // determine if a spike should occur
+        has_spiked = threshold_type_is_above_threshold(the_threshold_type);
```

Comment:
I suspect this is misnamed; it isn't really a threshold type any more, and it isn't a spike.

Reply:
inherited from before, but will change.

Reply:
fixed